### PR TITLE
paasta deployd watches for new watches when autoscaling is enabled first time

### PR DESCRIPTION
**Pull request by chlgit**
When the watcher is built for the first time, the event would be set as None according to kazoo document. (http://kazoo.readthedocs.io/en/latest/api/recipe/watchers.html) 

Test this change at mesosstage, and the log shows deployd createed service instance after a new node had been added into zookeeper successfully:

2018-01-31 13:37:10,706:mesos1-uswest1cdevc:DEBUG:paasta_tools.deployd.watchers.AutoscalerWatcher:Node change: data = b'2', stat = ZnodeStat(czxid=253550824220, mzxid=253550824221, ctime=1517434630666, mtime=1517434630669, version=1, cversion=0, aversion=0, ephemeralOwner=0, dataLength=1, numChildren=0, pzxid=253550824220), event = None
2018-01-31 13:37:10,707:mesos1-uswest1cdevc:INFO:paasta_tools.deployd.watchers.AutoscalerWatcher:Number of instances changed or autoscaling enabled for first time for robj-test-service.main_1
2018-01-31 13:37:10,761:mesos1-uswest1cdevc:DEBUG:paasta_tools.deployd.common.PaastaQueue:Adding ServiceInstance(service='robj-test-service', instance='main_1', bounce_by=1517434630, watcher='AutoscalerWatcher', bounce_timers=None, failures=0, priority=0) to InboxQueue queue

### CODE REVIEW

File: paasta_tools/deployd/watchers.py [hunk 1]
```diff
@@ -61,27 +61,31 @@ def watch_folder(self, path):
             for child in children:
                 self.watch_folder("{}/{}".format(path, child))
 
+    def _make_service_instance(self, path):
+        service, instance = path.split('/')[-3:-1]
```
Comment (solarkennedy) [inline paasta_tools/deployd/watchers.py]: This sounds

File: paasta_tools/deployd/watchers.py [hunk 2]
```diff
@@ -61,27 +61,31 @@ def watch_folder(self, path):
...
+            watcher=type(self).__name__,
+            failures=0,
+        )
+        self.inbox_q.put(service_instance)
+
     def watch_node(self, path):
-        self.log.debug("Adding watch on {}".format(path))
+        self.log.debug("Adding node watch on {}".format(path))
         DataWatch(self.zk, path, func=self.process_node_event, send_event=True)
+        self._make_service_instance(path)
```
Comment (solarkennedy) [inline paasta_tools/deployd/watchers.py]: This sounds OK, but is won't this pick up on all random node watches, like changes to files that are not relevant?

Specifically, don't we only want to call this when the nodepath ends in "instances" ?

**Comment by mattmb** [comment]
Thanks for taking a look at this, it's really appreciated!

When you were testing on mesosstage did you find that it *never* works without this patch? My understanding was that in general the code as it currently exists normally will add a watch on the newly created autoscaling nodes. But that sometimes something goes wrong and it "misses" creating the watch?

Another question: does the watcher get created but just not actioned on its first run (because of the event=None)? If so I think your fix explains it. But if you let the autoscaler create an `instances` key and change it and it still doesn't work then I'm surprised...

File: paasta_tools/deployd/watchers.py (hunk 2 above)
Comment (mattmb) [reply paasta_tools/deployd/watchers.py]: Should be okay because we only call this from a block that does an if on "instances"

File: paasta_tools/deployd/watchers.py [hunk 3]
```diff
@@ -61,27 +61,31 @@ def watch_folder(self, path):
             for child in children:
                 self.watch_folder("{}/{}".format(path, child))
 
+    def _make_service_instance(self, path):
```
Comment (mattmb) [inline paasta_tools/deployd/watchers.py]: Can you please call this `_enqueue_service_instance`

File: paasta_tools/deployd/watchers.py (hunk 2 above)
Comment (chlgit) [reply paasta_tools/deployd/watchers.py]: I think the check has been done at line 58:  " if children and ('instances' in children) "

File: paasta_tools/deployd/watchers.py (hunk 3 above)
Comment (chlgit) [reply paasta_tools/deployd/watchers.py]: Sure, will do

Comment (chlgit) [reply paasta_tools/deployd/watchers.py]: I saw the watcher being built without this patch, and it called process_node_event. But since event=None, it doesn't pick up when autoscaling is enabled first time. For instances been changed case, autoscaler works fine even without this patch.

Comment (mattmb) [reply paasta_tools/deployd/watchers.py]: Gotcha, I think this makes a lot of sense, it is for sure a bug!

I wonder if there's any other bugs that have a similar effect. Let's hope not!

Thanks again for taking a look :-)
